File: System-Design/Circuit-Breaker/cb.cpp
#include <iostream>
#include <functional>
#include <chrono>

enum class State { CLOSED, OPEN, HALF_OPEN };

class CircuitBreaker {
private:
    State state = State::CLOSED;
    int failureCount = 0;
    int failureThreshold;
    long long cooldownMs;
    long long nextAttemptTime = 0;

    long long now() {
        return std::chrono::duration_cast<std::chrono::milliseconds>(
            std::chrono::system_clock::now().time_since_epoch()
        ).count();
    }

    void recordFailure() {
        failureCount++;
        if (failureCount >= failureThreshold) {
            state = State::OPEN;
            nextAttemptTime = now() + cooldownMs;
        }
    }

    void reset() {
        failureCount = 0;
        state = State::CLOSED;
    }

public:
    CircuitBreaker(int threshold, long long cooldown) 
        : failureThreshold(threshold), cooldownMs(cooldown) {}

    std::string execute(std::function<bool()> action) {
        if (state == State::OPEN) {
            if (now() < nextAttemptTime) return "FAST_FAIL";
            state = State::HALF_OPEN;
        }

        try {
            if (!action()) throw std::runtime_error("Failed");
            reset();
            return "SUCCESS";
        } catch (...) {
            recordFailure();
            return "FAIL";
        }
    }
};
int main() {
    CircuitBreaker cb(2, 100);
    auto alwaysFail = []() { return false; };
    
    cb.execute(alwaysFail);
    cb.execute(alwaysFail); 
    std::string res = cb.execute(alwaysFail); 
    
    if (res == "FAST_FAIL") {
        std::cout << "C++ Test Passed!\n";
        return 0;
    }
    return 1;
}

File: System-Design/Circuit-Breaker/cb.java
public class Main {
    enum State { CLOSED, OPEN, HALF_OPEN }

    static class CircuitBreaker {
        private State state = State.CLOSED;
        private int failureCount = 0;
        private final int failureThreshold;
        private final long cooldownMs;
        private long nextAttemptTime = 0;

        public CircuitBreaker(int threshold, long cooldownMs) {
            this.failureThreshold = threshold;
            this.cooldownMs = cooldownMs;
        }

        public String execute(java.util.function.Supplier<Boolean> action) {
            if (state == State.OPEN) {
                if (System.currentTimeMillis() < nextAttemptTime) {
                    return "FAST_FAIL";
                }
                state = State.HALF_OPEN;
            }

            try {
                boolean success = action.get();
                if (!success) throw new RuntimeException("Action failed");
                reset();
                return "SUCCESS";
            } catch (Exception e) {
                recordFailure();
                return "FAIL";
            }
        }

        private void recordFailure() {
            failureCount++;
            if (failureCount >= failureThreshold) {
                state = State.OPEN;
                nextAttemptTime = System.currentTimeMillis() + cooldownMs;
            }
        }

        private void reset() {
            failureCount = 0;
            state = State.CLOSED;
        }
    }
    public static void main(String[] args) {
        CircuitBreaker cb = new CircuitBreaker(2, 100);
        cb.execute(() -> false); 
        cb.execute(() -> false); 
        String res = cb.execute(() -> false); 
        
        if ("FAST_FAIL".equals(res)) {
            System.out.println("Java Test Passed!");
        } else {
            System.exit(1);
        }
    }
}